app/src/main/java/com/navapp/navigation/route/OptimizingRouteFactory.java

```diff
@@ -0,0 +1,62 @@
+package com.navapp.navigation.route;
+
+import com.google.maps.GeoApiContext;
+import com.google.maps.errors.ApiException;
+import com.navapp.navigation.destination.Destination;
+import com.navapp.navigation.destination.data.Location;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class OptimizingRouteFactory implements RouteFactory{
+    private final RoutePlanner routePlanner;
+
+    public OptimizingRouteFactory(GeoApiContext context) {
+        this(new RoutePlanner(context));
+    }
+
+    public OptimizingRouteFactory(RoutePlanner routePlanner) {
+        this.routePlanner = routePlanner;
+    }
+
+    @Override
+    public Route create(Location start, List<Destination> destinations, Location end) throws InterruptedException, ApiException, IOException {
+        List<String> destinationIds = extractIds(destinations);
+        List<String> orderedDestinationIds = routePlanner.orderedRouteIds(start.getId(), destinationIds, end.getId());
+        List<Destination> orderedDestinations = orderDestinationsByIds(destinations, orderedDestinationIds, end);
+
+        return new Route(orderedDestinations);
+    }
+
+    private List<String> extractIds(List<Destination> destinations) {
+        List<String> destinationIds = new ArrayList<>();
+
+        for (Destination destination: destinations)
+            destinationIds.add(destination.getLocationId());
+
+        return destinationIds;
+    }
+
+    private List<Destination> orderDestinationsByIds(List<Destination> destinations, List<String> orderedIds, Location end) {
+        List<Destination> orderedDestinations = new ArrayList<>();
+
+        for (int i = 0; i < orderedIds.size()-1; ++i)
+            orderedDestinations.add(correspondingDestination(destinations, orderedIds.get(i)));
+
+        orderedDestinations.add(new Destination(end));
+
+        return orderedDestinations;
+    }
+
+    private Destination correspondingDestination(List<Destination> destinations, String id) {
+        for (int i = 0; i < destinations.size(); ++i)
+            if(isEqualId(destinations.get(i), id))
+                return destinations.remove(i);
+        return null;
```

Comment:
Maybe do some better handling then having it return `null` into the `List`?

Reply:
It shouldn't happen in the use case, but there is no good way to do so... I can make a dummy Destination, but then it will have a null Location or a dummy address which will also be bad

Reply:
So I would probably throw an Exception